File: go/0090M-子集II.go
/*
 * @Author: SukiEva
 * @Date: 2021-11-15 08:57:00
 * @Description: https://github.com/SukiEva
 * わたし、気になります！
 */
package main

import "sort"

func subsetsWithDup(nums []int) [][]int {
	ans := make([][]int, 0)
	sub := make([]int, 0)
	var dfs func(index int)
	dfs = func(index int) {
		tmp := make([]int, len(sub))
		copy(tmp, sub)
		ans = append(ans, tmp)
		if index >= len(nums) {
			return
		}
		for i := index; i < len(nums); i++ {
			if i > index && nums[i] == nums[i-1] {
				continue
			}
			sub = append(sub, nums[i])
			dfs(i + 1)
			sub = sub[:len(sub)-1]
		}
	}
	sort.Ints(nums)
	dfs(0)
	return ans
}

File: cpp/0090M-子集II.cpp
/*
 * @Author: SukiEva
 * @Date: 2021-12-27 15:37:43
 * @Description: https://github.com/SukiEva
 * わたし、気になります！
 */
#include <bits/stdc++.h>

#include "include/ListNode.h"
#include "include/TreeNode.h"

using namespace std;

class Solution {
   public:
    vector<vector<int>> ans;
    vector<int> tmp;

    void dfs(vector<int>& nums, int idx) {
        ans.push_back(tmp);
        if (idx >= nums.size()) return;
        for (int i = idx; i < nums.size(); i++) {
            if (i > idx && nums[i] == nums[i - 1]) continue;
            tmp.push_back(nums[i]);
            dfs(nums, i + 1);
            tmp.pop_back();
        }
    }

    vector<vector<int>> subsetsWithDup(vector<int>& nums) {
        sort(nums.begin(), nums.end());
        dfs(nums, 0);
        return ans;
    }
};